### ISSUE TITLE
Quitting while thread is adding CanvasItem-derived nodes creates a deadlock and hangs the application indefinitely

### ISSUE THREAD

**Issue by 2shady4u** [godotengine/godot#43466]
**Godot version:**
3.2.3

**OS/device including version:**
Windows 10

**Issue description:**
When using threads you should, as the docs indicate, always wait for them to finish so they can be joined to the main thread. Otherwise you get memory leakage and unexpected behavior.

When using the piece of minimal code and attempting to quit the game, the application gets stuck in an endless loop (for some reason?)

```Swift
extends Node2D

var _threads := []

func _ready():
	var thread = Thread.new()
	_threads.append(thread)
	var _error = thread.start(self, "_async_load", [thread])

func _async_load(argument_values : Array) -> Node2D:
	print("_async_load_completed")
	var thread : Thread = argument_values[0]

	var node : Node2D = Node2D.new()
	for i in 10000:
		print(i)
		var control : Control = Control.new()
		node.add_child(control)

	call_deferred("_async_load_completed", thread)
	return node

func _async_load_completed(thread : Thread) -> void:
	print("_async_load_completed")

	var node = thread.wait_to_finish()
	_threads.erase(thread)

	add_child(node)

func _exit_tree():
	print("_exit_tree")

	for thread in _threads:
		var node = thread.wait_to_finish()
		node.free()
```

**How to reproduce?**

Run the above piece of code on a Node2D and try to quit the application (= pressing the X button) before all 10.000 controls have been added. The application will hang indefinitely...

When waiting for all 10.000 controls to be added, the application quits correctly without issue.

This hanging indefinitely does NOT happen when the Control is replaced by a Timer, Tween or even Spatial (or **anything that does not derive from CanvasItem**)

In other words, this does not get stuck in an endless loop:

```Swift
for i in 10000:
	print(i)
	var tween : Tween = Tween.new()
	node.add_child(control)
```

Only adding nodes that inherit from CanvasItem seems to create this deadlock...
Maybe I'm just doing something wrong here? If so, please enlighten me, but I have no clue what could be wrong here...

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/7235?

**Comment by 2shady4u** [comment]
Small update/ more information:

This hanging indefinitely does NOT happen when the Control is replaced by a Timer or Tween (or anything that does not derive from CanvasItem)

In other words, this does not get stuck in an endless loop:

```
for i in 10000:
		print(i)
		var tween : Tween = Tween.new()
		node.add_child(control)
```

**Comment by 2shady4u** [comment]
@Calinou I doubt those issues are related. In #7235 the issue OP described relates to having to wait for thread completion which might hang the application temporarily (= a few seconds). This is however fixed by the """workaround""" described in that very thread (by both @bojidar-bg and @eska014).

In my case, the application will hang indefinitely, possible due to a deadlock of some kind. 
I already use this """workaround""" in some way or form, since I already wait for thread completion.

**Comment by 2shady4u** [comment]
**I found a crude, but functioning work-around:**

```Swift
extends Node2D

var _threads := []
var quit_on_thread_completion := false

func _ready():
	get_tree().set_auto_accept_quit(false)

	var thread = Thread.new()
	_threads.append(thread)
	var _error = thread.start(self, "_async_load", [thread])

func _notification(what):
	if what == NOTIFICATION_WM_QUIT_REQUEST:
		quit_on_thread_completion = true

func _async_load(argument_values : Array) -> Node2D:
	print("_async_load_completed")
	var thread : Thread = argument_values[0]

	var node : Node2D = Node2D.new()
	for i in 10000:
		print(i)
		var control : Control = Control.new()
		node.add_child(control)

	call_deferred("_async_load_completed", thread)
	return node

func _async_load_completed(thread : Thread) -> void:
	print("_async_load_completed")

	var node = thread.wait_to_finish()
	_threads.erase(thread)

	add_child(node)

	if quit_on_thread_completion:
		get_tree().quit()
	else:
		get_tree().set_auto_accept_quit(true)
```

The issue still persists however, as CanvasItem-derived nodes should not deadlock a thread like this!

**Comment by Zylann** [comment]
`CanvasItem` invokes this in its constructor:
```cpp
	canvas_item = VisualServer::get_singleton()->canvas_item_create();
```
One thing to check in your project: the renderer must be set to a thread-safe mode. If it isn't, undefined behavior will happen.

**Comment by 2shady4u** [comment]
@Zylann that seems to do the job actually...
Setting the rendering thread model to "multi-threading" stops the deadlock.

This **extremely important** setting is however NEVER mentioned in any of the docs.
I will open a new issue or PR on the docs repo to alleviate this shortcoming.

I'll leave this issue open for now until I make the change to the docs

**Comment by Calinou** [comment]
Keep in mind the multi-threaded rendering thread model can break some things, so you need to use it consciously.

**Comment by Zylann** [comment]
Also, I wouldn't expect it to break if it is set  to `Single-Safe`, which is actually the default.

**Comment by 2shady4u** [comment]
@Zylann It does however deadlock when the threading model is set to "Single-Safe"...

**Comment by Calinou** [comment]
@2shady4u Can you (or anyone else) still reproduce this bug in [Godot 3.3](https://godotengine.org/article/godot-3-3-has-arrived) or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by jitspoe** [comment]
Same issue as https://github.com/godotengine/godot/issues/39392?

**Comment by Calinou** [comment]
Closing per @jitspoe's comment.

**Comment by Calinou** [comment]
Closing per @jitspoe's comment.
